_posts/2023-03-27-group-equivariant-cnn-2/group-equivariant-cnn-2.Rmd

```diff
@@ -0,0 +1,374 @@
+
+First, we prepare the data; in particular, the augmented test set.
+
+```{r}
+dir <- "/tmp/mnist"
+
+train_ds <- torchvision::mnist_dataset(
+  dir,
+  download = TRUE,
+  transform = torchvision::transform_to_tensor
+)
+
+test_ds <- torchvision::mnist_dataset(
+  dir,
+  train = FALSE,
+  transform = function(x) {
+    x |>
+      torchvision::transform_to_tensor() |>
+      torchvision::transform_random_rotation(
+        degrees = c(0, 360),
+        resample = 2,
+        fill = 0
+      )
+  }
+)
+
+train_dl <- dataloader(train_ds, batch_size = 128, shuffle = TRUE)
+test_dl <- dataloader(test_ds, batch_size = 128)
+```
+
+How does it look?
+
+```{r}
+test_images <- coro::collect(
+  test_dl, 1
+)[[1]]$x[1:32, 1, , ] |> as.array()
+
+par(mfrow = c(4, 8), mar = rep(0, 4), mai = rep(0, 4))
+test_images |>
+  purrr::array_tree(1) |>
+  purrr::map(as.raster) |>
+  purrr::iwalk(~ {
+    plot(.x)
+  })
+```
+
+```{r, echo=FALSE, eval=TRUE, fig.alt = "32 digits, rotated randomly."}
+knitr::include_graphics("images/rotated-mnist.png")
+```
+
+We first define and train a conventional CNN. It is as similar to `GroupEquivariantCNN()`, architecture-wise, as possible, and is given twice the number of hidden channels, so as to have comparable capacity overall.
+
+```{r}
+ default_cnn <- nn_module(
+   "default_cnn",
+   initialize = function(kernel_size, in_channels, out_channels, num_hidden, hidden_channels) {
+     self$conv1 <- torch::nn_conv2d(in_channels, hidden_channels, kernel_size)
+     self$convs <- torch::nn_module_list()
+     for (i in 1:num_hidden) {
+       self$convs$append(torch::nn_conv2d(hidden_channels, hidden_channels, kernel_size))
+     }
+     self$avg_pool <- torch::nn_adaptive_avg_pool2d(1)
+     self$final_linear <- torch::nn_linear(hidden_channels, out_channels)
+   },
+   forward = function(x) {
+     x <- x |>
+       self$conv1() |>
+       (\(.) torch::nnf_layer_norm(., .$shape[2:4]))() |>
+       torch::nnf_relu()
+     for (i in 1:(length(self$convs))) {
+       x <- x |>
+         self$convs[[i]]() |>
+         (\(.) torch::nnf_layer_norm(., .$shape[2:4]))() |>
+         torch::nnf_relu()
+     }
+     x <- x |>
+       self$avg_pool() |>
+       torch::torch_squeeze() |>
+       self$final_linear()
+     x
+   }
+ )
+
+fitted <- default_cnn |>
+    luz::setup(
+      loss = torch::nn_cross_entropy_loss(),
+      optimizer = torch::optim_adam,
+      metrics = list(
+        luz::luz_metric_accuracy()
+      )
+    ) |>
+    luz::set_hparams(
+      kernel_size = 5,
+      in_channels = 1,
+      out_channels = 10,
+      num_hidden = 4,
+      hidden_channels = 32
+    ) %>%
+    luz::set_opt_hparams(lr = 1e-2, weight_decay = 1e-4) |>
+    luz::fit(train_dl, epochs = 10, valid_data = test_dl) 
+
+```
+
+    Train metrics: Loss: 0.0498 - Acc: 0.9843
+    Valid metrics: Loss: 3.2445 - Acc: 0.4479
+
+Unsurprisingly, accuracy on the test set is not that great.
```

Comment:
Unsurprisingly why? Unsurprisingly…without group equivariance? (Ignore if this is obvious to someone who's a more informed reader, which isn't a terribly high bar 😉 )

Reply:
Yeah ... you're right. But well, thinking of the typical audience ... I agree, I could leave it in I guess